sdk/metric/meter.go

```diff
@@ -201,28 +208,115 @@ func (m *meter) Float64ObservableGauge(name string, options ...instrument.Float6
 // RegisterCallback registers the function f to be called when any of the
 // insts Collect method is called.
 func (m *meter) RegisterCallback(f metric.Callback, insts ...instrument.Asynchronous) (metric.Registration, error) {
+	if len(insts) == 0 {
+		// Don't allocate a observer if not needed.
+		return noopRegister{}, nil
+	}
+
+	reg := newObserver()
+	var errs multierror
 	for _, inst := range insts {
-		// Only register if at least one instrument has a non-drop aggregation.
-		// Otherwise, calling f during collection will be wasted computation.
-		switch t := inst.(type) {
-		case *instrumentImpl[int64]:
-			if len(t.aggregators) > 0 {
-				return m.registerMultiCallback(f)
+		switch o := inst.(type) {
+		case *observable[int64]:
+			if err := o.registerable(m.scope); err != nil {
+				if !errors.Is(err, errEmptyAgg) {
+					errs.append(err)
+				}
+				continue
 			}
-		case *instrumentImpl[float64]:
-			if len(t.aggregators) > 0 {
-				return m.registerMultiCallback(f)
+			reg.registerInt64(o.observablID)
+		case *observable[float64]:
+			if err := o.registerable(m.scope); err != nil {
+				if !errors.Is(err, errEmptyAgg) {
+					errs.append(err)
+				}
+				continue
 			}
+			reg.registerFloat64(o.observablID)
 		default:
-			// Instrument external to the SDK. For example, an instrument from
-			// the "go.opentelemetry.io/otel/metric/internal/global" package.
-			//
-			// Fail gracefully here, assume a valid instrument.
-			return m.registerMultiCallback(f)
+			// Instrument external to the SDK.
+			return nil, fmt.Errorf("invalid observable: from different implementation")
 		}
 	}
-	// All insts use drop aggregation.
-	return noopRegister{}, nil
+
+	if err := errs.errorOrNil(); err != nil {
+		return nil, err
+	}
+
+	if reg.len() == 0 {
+		// All insts use drop aggregation.
+		return noopRegister{}, nil
+	}
+
+	cback := func(ctx context.Context) error {
+		return f(ctx, reg)
+	}
+	return m.pipes.registerMultiCallback(cback), nil
+}
+
+type observer struct {
+	float64 map[observablID[float64]]struct{}
+	int64   map[observablID[int64]]struct{}
+}
+
+func newObserver() observer {
+	return observer{
+		float64: make(map[observablID[float64]]struct{}),
+		int64:   make(map[observablID[int64]]struct{}),
+	}
+}
+
+func (r observer) len() int {
+	return len(r.float64) + len(r.int64)
+}
+
+func (r observer) registerFloat64(id observablID[float64]) {
+	r.float64[id] = struct{}{}
+}
+
+func (r observer) registerInt64(id observablID[int64]) {
+	r.int64[id] = struct{}{}
+}
+
+var (
+	errUnknownObserver = errors.New("unknown observable")
+	errUnregObserver   = errors.New("observable not registered for callback")
```

Comment:
```suggestion
	errUnregObserver   = errors.New("instrument not registered with callback")
```

Minor change.

Reply:
`observable` here helps communicate to the user that it is an observable instrument that the error applies to. I would perfer to be specific in the error message to help communicate this.

Reply:
`observable` is an unexposed type, users interact with instruments.  What we are looking to tell the user is that the instrument they are trying to use with the Observer isn't registered in this callback.

Could we split the difference and use `observable instrument`?

Reply:
> Could we split the difference and use `observable instrument`?

Sure :+1: